File: NodeLink/92. 反转链表 II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
 var reverseBetween = function(head, left, right) {
    let node = new ListNode(-1,head);
    let left1,right1;
    left1 = node;

    right1 = node;
    for(let i = 0 ; i < left - 1;i++){
        left1 = left1.next; 
    }
    for(let i = 0 ; i < right + 1;i++){
        right1 = right1.next;
    }

    let reverseEnd = left1.next;
    let reverseStrat = reverseN(reverseEnd,right - left + 1);

    left1.next = reverseStrat;
    reverseEnd.next = right1;
    return node.next;
};

var reverseN = function (head,n){
    let pre , cur;
    pre = null;
    cur = head;
    for(let i = 0;i < n ; i++){
        let info = cur.next;
        cur.next = pre;
        pre = cur;
        cur = info;
    }
    return pre;
}

File: Python/链表反转/92. 反转链表 II.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-6 9:54
# @Author : 刘永奇
# @File : 92. 反转链表 II.py
# @Software : PyCharm
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head or not head.next:return head
        # 小技巧：使用虚拟头节点来兼容从头开始反转的情况
        h = ListNode(0,head)
        p  = h
        m = left
        while m > 1:
            p = p.next
            m -= 1
        p.next = self.reverse(p.next,right-left+1)
        return h.next

    def reverse(self,head,n):
        if n==1:return head
        tail = head.next
        p = self.reverse(head.next,n-1)
        head.next = tail.next
        tail.next = head
        return p